### PR TITLE
Users Change email feature

### LINKED ISSUE #1630: More user functionalities
As a user besides changing my password I would like to:

1. Change my e-mail address.
2. Delete my account.
3. Create a new password if I forgot my password.

### PR DESCRIPTION
**Pull request by agriffard**
Adds a feature and settings to allow the users to change their email.

Fixes #1630

### CODE REVIEW

**Comment by Skrypt** [comment]
Should you also make this optional on the User Edit form?

image

**Comment by agriffard** [comment]
It was only to allow it on the front end.
I could do the same in the Edit user in admin (retrieve the value of the setting and display the email instead of an input if enabled) but do we want this beahviour?

**Comment by Skrypt** [comment]
Merge it then. If there is a request for the admin UI later on it's an easy fix.

File: src/OrchardCore.Modules/OrchardCore.Users/Controllers/ChangeEmailController.cs [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+                var user = await _userService.GetAuthenticatedUserAsync(User);
+                var userWithEmail = await _userManager.FindByEmailAsync(model.Email);
+
+                if (((User)user).Email == model.Email)
+                {
+                    ModelState.AddModelError("Email", S["This email is already your current one."]);
+                }
+                else if (userWithEmail != null && user.UserName != userWithEmail.UserName)
+                {
+                    ModelState.AddModelError("Email", S["A user with the same email already exists."]);
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Users/Controllers/ChangeEmailController.cs]: Return the same error as when a new account with an existing email is created.

File: src/OrchardCore.Modules/OrchardCore.Users/Controllers/ChangeEmailController.cs [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+            {
+                return NotFound();
+            }
+
+            if (ModelState.IsValid)
+            {
+                var user = await _userService.GetAuthenticatedUserAsync(User);
+                var userWithEmail = await _userManager.FindByEmailAsync(model.Email);
+
+                if (((User)user).Email == model.Email)
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Users/Controllers/ChangeEmailController.cs]: Case insensitive copmarison

File: src/OrchardCore.Modules/OrchardCore.Users/Models/ChangeEmailViewModel.cs [hunk 3]
```diff
@@ -0,0 +1,11 @@
+using System.ComponentModel.DataAnnotations;
+
+namespace OrchardCore.Users.ViewModels
+{
+    public class ChangeEmailViewModel
+    {
+        [Required]
+        [DataType(DataType.EmailAddress)]
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Users/Models/ChangeEmailViewModel.cs]: Use `IValidatableObject` and MailKit

File: src/OrchardCore.Modules/OrchardCore.Users/Controllers/ChangeEmailController.cs [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+        [HttpGet]
+        public async Task<IActionResult> Index()
+        {
+            if (!(await _siteService.GetSiteSettingsAsync()).As<ChangeEmailSettings>().AllowChangeEmail)
+            {
+                return NotFound();
+            }
+
+            var user = await _userService.GetAuthenticatedUserAsync(User);
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Users/Controllers/ChangeEmailController.cs]: Check user is actually authenticated. Or return Challenge.

Comment (sebastienros) [reply src/OrchardCore.Modules/OrchardCore.Users/Controllers/ChangeEmailController.cs]: Or add [Authorize] on the controller

File: src/OrchardCore.Modules/OrchardCore.Users/Controllers/ChangeEmailController.cs (hunk 1 above)
Comment (agriffard) [reply src/OrchardCore.Modules/OrchardCore.Users/Controllers/ChangeEmailController.cs]: Do you mean 'Add the same email check in the RegistrationController'?

**Pull request merged**